store/src/java-test/com/zimbra/cs/event/analytics/EventMetricTest.java

```diff
@@ -0,0 +1,270 @@
+        EventMetric<RatioMetric, Double, RatioIncrement> contactReadRatio = metrics.getMetric(key);
+
+        EventDifferenceParams globalParams = new EventDifferenceParams(EventType.READ, EventType.SEEN);
+        globalParams.setInitializer(initializer);
+        key = new MetricKey<RatioMetric, Double, RatioIncrement>(MetricType.EVENT_RATIO, globalParams);
+        EventMetric<RatioMetric, Double, RatioIncrement> globalReadRatio = metrics.getMetric(key);
+
+        assertTrue("initializer.getInitialValue() should have been triggered", initializedFlag.get() == true);
+        assertEquals("initial contact1 read ratio should be 0", new Double(0), contactReadRatio.getValue());
+        assertEquals("initial global read ratio should be 0", new Double(0), globalReadRatio.getValue());
+
+        logMsgEvents(CONTACT_1, 1, 3, EventType.SEEN);
+        logMsgEvents(CONTACT_1, 1, 2, EventType.READ);
+        logMsgEvents(CONTACT_2, 10, 3, EventType.SEEN);
+        logMsgEvents(CONTACT_2, 10, 1, EventType.READ);
+
+        assertEquals("new contact1 read ratio should be 2/3", new Double(2d/3), contactReadRatio.getValue());
+        assertEquals("new global read ratio should be 1/2", new Double(1d/2), globalReadRatio.getValue());
+    }
+
+    @Test
+    public void testTimeToOpen() throws Exception {
+        AtomicBoolean initializedFlag = new AtomicBoolean(false);
+        MetricInitializer<RatioMetric, Double, RatioIncrement> initializer = new DummyRatioInitializer(initializedFlag);
+
+        EventDifferenceParams contactParams = new EventDifferenceParams(EventType.SEEN, EventType.READ, CONTACT_1);
+        contactParams.setInitializer(initializer);
+        MetricKey<RatioMetric, Double, RatioIncrement> key = new MetricKey<RatioMetric, Double, RatioIncrement>(MetricType.TIME_DELTA, contactParams);
+        EventMetric<RatioMetric, Double, RatioIncrement> contactTTO = metrics.getMetric(key);
+
+        EventDifferenceParams globalParams = new EventDifferenceParams(EventType.SEEN, EventType.READ);
+        globalParams.setInitializer(initializer);
+        key = new MetricKey<RatioMetric, Double, RatioIncrement>(MetricType.TIME_DELTA, globalParams);
+        EventMetric<RatioMetric, Double, RatioIncrement> globalTTO = metrics.getMetric(key);
+
+        assertTrue("initializer.getInitialValue() should have been triggered", initializedFlag.get() == true);
+        assertEquals("initial contact1 TTO should be 0", new Double(0), contactTTO.getValue());
+        assertEquals("initial global TTO should be 0", new Double(0), globalTTO.getValue());
+
+        long timestamp = System.currentTimeMillis();
+        //1 second difference for contact1
+        logMsgEvents(CONTACT_1, 1, 1, EventType.SEEN, timestamp-1000);
+        logMsgEvents(CONTACT_1, 1, 1, EventType.READ, timestamp);
+
+        //2 second difference for contact2
```

Comment:
why do you need to log events for CONTACT_2 if you are not checking them in this test?

Reply:
Logging events for CONTACT_2 changes the global time-to-open to 1.5 seconds.
Although I just realized there's a copy/paste error in the error message below this.